### PR TITLE
Add edit commands

### PR DESCRIPTION
**Pull request by bentanjunrong**
Lets:

1. Add all the edit command classes and relevant methods in ModBook, Model/ModelManager etc.
1. Add tests for all added edit commands
1. Add utility methods to construct command string for given Edit Mod/Exam/Lesson Descriptors

Fix for #89.
1. Remove c/CODE prefix from edit lesson and exam commands since there is no ambiguity in the commands due to only mod being shown when the command is entered.
1. Update Command Usage for Edit/Delete Exam/Lesson Commands.
1. Update tests with new command syntax.
1. Update UG with new command syntax.

### CODE REVIEW

**Comment by bentanjunrong** [comment]
> Not sure why `execute_duplicateExam_throwsCommandException` and `execute_newExam_success` of `AddCommandIntegrationTest` are now failing. Didn't touch AddCommand.

Fixed, accidentally made some changes in ModelManager that broke AddExamCommands.

File: src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        Exam examToEdit = exams.get(targetIndex.getZeroBased());
+        Exam editedExam = createEditedExam(examToEdit, editExamDescriptor);
+
+        if (!examToEdit.isSameExam(editedExam) && model.moduleHasExam(module, editedExam)) {
+            throw new CommandException(MESSAGE_DUPLICATE_EXAM);
+        }
+
+        model.setExam(module, examToEdit, editedExam);
+        model.updateFilteredModuleList(PREDICATE_SHOW_ALL_MODULES);
+        return new CommandResult(String.format(MESSAGE_EDIT_EXAM_SUCCESS, editedExam, module));
```
Comment (luminousleek) [inline src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java:87]: wondering if editing a lesson / exam should make modbook stay in `GuiState.DETAILS`. this means updating the `FilteredModuleList` wouldn't be necessary

File: src/main/java/seedu/address/model/ModBook.java [hunk 2]
```diff
@@ -110,6 +110,24 @@ public void removeLesson(Module module, Lesson target) {
         module.getLessons().remove(target);
     }
 
+    /**
+     * Sets Exam {@code target} in the Module {@code module}.
+     * {@code target} must exist in {@code module}'s exams list.
+     */
+    public void setExam(Module module, Exam target, Exam newExam) {
+        List<Exam> examList = module.getExams();
```
Comment (luminousleek) [inline src/main/java/seedu/address/model/ModBook.java:118]: maybe to satisfy law of demeter (i.e. don't talk to strangers), can implement `Module::setExam` and `Module::setLesson` methods?

if you do this then you'll probably also have to go and edit the delete/add lesson/exam methods in `ModelManager`, shown below:
image
like right now there are 2 different ways of doing similar things, and yours looks more like the remove methods. should standardise la.

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 3]
```diff
@@ -117,7 +117,7 @@
      * - the {@code actualModel} matches {@code expectedModel}
      */
     public static void assertCommandSuccess(Command command, Model actualModel, CommandResult expectedCommandResult,
-            Model expectedModel) {
+                                            Model expectedModel) {
```
Comment (luminousleek) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java:120]: didn't know this passes checkstyle 😮

File: src/test/java/seedu/address/logic/commands/edit/EditExamCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
+package seedu.address.logic.commands.edit;
```
Comment (luminousleek) [inline src/test/java/seedu/address/logic/commands/edit/EditExamCommandTest.java:1]: 😢 all the best sir

File: src/test/java/seedu/address/logic/commands/edit/EditExamCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    public void equals() {
+
+    }
+
+    /**
+     * Updates {@code model}'s filtered list to show no one.
+     */
+    private void showNoModule(Model model) {
+        model.updateFilteredModuleList(m -> false);
+        assertTrue(model.getFilteredModuleList().isEmpty());
```
Comment (luminousleek) [inline src/test/java/seedu/address/logic/commands/edit/EditExamCommandTest.java:87]: pretty sure this method appears somewhere else? maybe can shift it to some testutil class

File: src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
```
Comment (SHEZADHASSAN22) [inline src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java:87]: This should bbe editedExam.getName and module.getCode

Comment (bentanjunrong) [reply src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java]: Thank you for your incredible insight, Professor Hassan22

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java (hunk 3 above)
Comment (bentanjunrong) [reply src/test/java/seedu/address/logic/commands/CommandTestUtil.java:120]: Haha configured my IntelliJ to FormatOnSave according to our checkstyle standards so there's some random changes like this

File: src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+            + PREFIX_VENUE + "Field";
+    private final Index targetIndex;
+    private final ModuleCode targetModuleCode;
+    private final EditExamDescriptor editExamDescriptor;
+
+    /**
+     * Creates an EditExamCommand to edit the Exam at specified {@code Index} of the specified {@code Module}.
+     */
+    public EditExamCommand(Index targetIndex, ModuleCode targetModuleCode, EditExamDescriptor editExamDescriptor) {
+        requireAllNonNull(targetIndex, targetModuleCode, editExamDescriptor);
```
Comment (luminousleek) [inline src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java:60]: shucks i'm reviewing rohit's pr about removing code from delete exam/lesson (issue #89) and i think you can also remove it since edit exam/lesson is only done in the detail view.

but if you want to do another pr at another time then that's fine too

Comment (bentanjunrong) [reply src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java]: No prob, fixed it in this one

File: docs/UserGuide.md [hunk 8]
```diff
@@ -1,174 +1,222 @@
...
-3. [FAQ](#FAQ)
-4. [Date and time formats](#Date-and-time-formats)
-5. [Command Summary](#Command-Summary)
-    * [By Action](#Command-Summary-by-Action)
-    * [By Object](#Command-Summary-by-Object)
 
+- [User Guide](#user-guide)
+  - [Table of Contents](#table-of-contents)
+  - [Quick Start](#quick-start)
+  - [Features](#features)
```
Comment (rohit0718) [inline docs/UserGuide.md:12]: why change to bullets?

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -2,64 +2,66 @@
 layout: page
 title: Developer Guide
 ---
-* Table of Contents
-{:toc}
 
---------------------------------------------------------------------------------------------------------------------
+- Table of Contents
+  {:toc}
+
```
Comment (rohit0718) [inline docs/DeveloperGuide.md:8]: what is this space and bullet point change for?

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -2,64 +2,66 @@
...
 
---------------------------------------------------------------------------------------------------------------------
+---
 
 ## **Setting up, getting started**
 
 Refer to the guide [_Setting up and getting started_](SettingUp.md).
 
---------------------------------------------------------------------------------------------------------------------
+---
```
Comment (rohit0718) [inline docs/DeveloperGuide.md:21]: why change these as well?

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -2,64 +2,66 @@
...
 :bulb: **Tip:** The `.puml` files used to create diagrams in this document can be found in the [diagrams](https://github.com/se-edu/addressbook-level3/tree/master/docs/diagrams/) folder. Refer to the [_PlantUML Tutorial_ at se-edu/guides](https://se-education.org/guides/tutorials/plantUml.html) to learn how to create and edit diagrams.
+
 </div>
 
 ### Architecture
 
 <img src="images/ArchitectureDiagram.png" width="280" />
 
-The ***Architecture Diagram*** given above explains the high-level design of the App.
+The **_Architecture Diagram_** given above explains the high-level design of the App.
```
Comment (rohit0718) [inline docs/DeveloperGuide.md:35]: any particular reason for this italics change here and everywhere else?

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -341,9 +347,10 @@ Use case ends.
 &nbsp;&nbsp;&nbsp; 1a2. User enters new data  
 &nbsp;&nbsp;&nbsp; 1a3. Use case resumes from step 1
 
-**Use case: Add a lesson**
```
Comment (rohit0718) [inline docs/DeveloperGuide.md:344]: are u sure this renders properly if u remove this line? @luminousleek had an issue where u need to put the space for it to go to the next line

**Review by rohit0718 — CHANGES_REQUESTED** [review]
jus reviewed the docs part.

File: src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+import seedu.address.model.module.exam.Exam;
+import seedu.address.model.module.exam.ExamName;
+
+public class EditExamCommand extends EditCommand {
+    public static final String MESSAGE_DUPLICATE_EXAM = "This exam already exists.";
+    public static final String MESSAGE_EDIT_EXAM_SUCCESS = "Edited Exam %s in Module %s";
+    public static final String MESSAGE_NOT_EDITED = "At least one field to edit must be provided.";
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Edits an exam in the Mod book. "
+            + "\nParameters: INDEX "
+            + PREFIX_CODE + "MOD_CODE "
```
Comment (luminousleek) [inline src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java:37]: now doesn't need code right?

File: src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+            + PREFIX_NAME + "EXAM_NAME "
+            + PREFIX_DAY + "DAY "
+            + PREFIX_START + "START_TIME "
+            + PREFIX_END + "END_TIME "
+            + PREFIX_LINK + "LINK "
+            + PREFIX_VENUE + "VENUE "
+            + "\nExample: " + COMMAND_WORD + " exam 1 "
+            + PREFIX_CODE + "CS2103 "
```
Comment (luminousleek) [inline src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java:45]: remove this too

File: src/main/java/seedu/address/logic/commands/edit/EditLessonCommand.java [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+import seedu.address.model.module.lesson.Lesson;
+import seedu.address.model.module.lesson.LessonName;
+
+public class EditLessonCommand extends EditCommand {
+    public static final String MESSAGE_DUPLICATE_LESSON = "This lesson already exists.";
+    public static final String MESSAGE_EDIT_LESSON_SUCCESS = "Edited Lesson %s in Module %s";
+    public static final String MESSAGE_NOT_EDITED = "At least one field to edit must be provided.";
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Edits a lesson in the Mod book. "
+            + "\nParameters: INDEX "
+            + PREFIX_CODE + "MOD_CODE "
```
Comment (luminousleek) [inline src/main/java/seedu/address/logic/commands/edit/EditLessonCommand.java:37]: same thing here, no need for code

File: src/main/java/seedu/address/logic/parser/EditCommandParser.java [hunk 16]
```diff
@@ -1,83 +1,181 @@
...
+            throw new GuiStateException(MESSAGE_WRONG_VIEW_DETAILS);
         }
-        if (argMultimap.getValue(PREFIX_EMAIL).isPresent()) {
-            editPersonDescriptor.setEmail(ParserUtil.parseEmail(argMultimap.getValue(PREFIX_EMAIL).get()));
+
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_NAME, PREFIX_DAY,
+                PREFIX_START, PREFIX_END, PREFIX_LINK, PREFIX_VENUE);
+        Index index = ParserUtil.parseFirstIndex(args);
+
+
```
Comment (luminousleek) [inline src/main/java/seedu/address/logic/parser/EditCommandParser.java:104]: really maximising the number of lines changed on reposense ah

**Review by luminousleek — CHANGES_REQUESTED** [review]
small changes left, thanks sir!!

File: src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+     * corresponding field value of the exam.
+     */
+    public static class EditExamDescriptor {
+        private ExamName name;
+        private ModBookDate date;
+        private Timeslot timeslot;
+        private Venue venue;
+        private Link link;
+
+        public EditExamDescriptor() {
```
[the remark this answers is not in the document]
Comment (bentanjunrong) [reply src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java:139]: Nope can't delete, I'm actually using it. Can go to the respective tests to see.

File: src/main/java/seedu/address/logic/commands/edit/EditLessonCommand.java [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+        private Timeslot timeslot;
+        private Venue venue;
+        private Link link;
+
+        public EditLessonDescriptor() {
+        }
+
+        /**
+         * Copy constructor.
+         * A defensive copy of {@code tags} is used internally.
```
[the remark this answers is not in the document]
Comment (bentanjunrong) [reply src/main/java/seedu/address/logic/commands/edit/EditLessonCommand.java:145]: thnx, removed.

File: src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java (hunk 13 above)
Comment (bentanjunrong) [reply src/main/java/seedu/address/logic/commands/edit/EditExamCommand.java:37]: Good catch. I realised I was referencing rohit's PR and his didn't change either ah. I'll make the changes for his files too

**Review by luminousleek — APPROVED** [review]
lgtm! except user guide which @rohit0718 will comment on 🎉 🥳 😃 ❤️ 😘

**Review by rohit0718 — APPROVED** [review]
lgtm! we have delayed this long enough alr, lets merge it. any ug dg changes lets settle later. thank u sir :')
